README.md
# YIcon Builder

如果希望在自己的服务器上搭建字体站，可以遵照如下的方式，一步一步进行部署即可。

## 环境要求

系统要求：

- Linux/Unix

环境要求：

- node 4.x+
- npm 3.x
- mysql 5.5+

npm 应该升级到 3.x，升级方式为：

```
$ sudo npm install -g npm@3 --registry=https://registry.npm.taobao.org
```

应该提前配置好 mysql 的服务，创建一个用户及其登录密码（可参考[这篇文章](http://www.cyberciti.biz/faq/mysql-change-root-password/)），并在里面创建一个 `utf-8` 编码的数据库。数据库名字、用户名、密码应和下面配置项中的名字保持一致。

## 初始安装

使用离线安装的方式，找到我们的压缩包 `yicon-builder.tar.gz`，运行命令：

```bash
$ npm install yicon-builder -g
```

安装完成之后，会在拥有一个 `yicon` 的命令。运行 `yicon init`，可以按照提示进行安装：

### 提示：输入安装路径

这个是项目的安装路径，安装时会生成如下的目录结构：

```
├── config.json
├── logs
└── src
```

### 提示：输入数据库配置项

这里是连接数据库的必备选项，包含以下内容：

1. 数据库域名，即数据库服务的域，如果配置在本地就填写默认 `127.0.0.1`；
2. 数据库用户名，即登录数据库的默认用户，默认为 `root`；
3. 数据库密码，必须按照上文所述，给数据库设定一个密码，默认为 `123456`；
4. 数据库端口号，默认 `3306`；
5. 数据库名称，必须在设置之前创建并指定一个特定的数据库，推荐为默认值 `iconfont`；

配置完成后将生成配置文件，并尝试连接数据库，如果连接期间存在问题，将无法自动导入数据表，安装程序会将创建数据表的 sql 文件置入安装路径中，需要稍后手动导入。

### 提示：选择登陆类型

由于本应用没有接入第三方登录，因此依赖内部系统的 cas 或 sso 登录模式。应用内部已经支持这两种模式的配置，只需要提供相关 url 即可。

在生成的配置文件中，我们会看到如下配置：

```javascript
{
    "login": {
        "ssoType": "cas",
        "authUrl": "http://cas.example.com/cas/login?service={{service}}",
        "tokenUrl": "http://cas.example.com/serviceValidate?service={{service}}&ticket={{token}}",
        "serviceUrl": "http://app.iconfont.com",
        "adminList": []
    }
}
```

这里面的配置项含义如下：

- `ssoType`: 根据用户输入自动生成登录类型，会分别按照 cas 和 sso 的规则进行登录；
- `authUrl`: 指点击登录按钮时需要跳转的 url，通常是一个登录页面，用户在这个页面登录后，会给服务端返回一个 token；
- `tokenUrl`: 服务端在获取 token 之后，可以通过这个 url 来获取用户的详细信息；
- `serviceUrl`: 通常这两种服务都需要提供当前域名，这里可以配置 iconfont 字体站服务的域名；
- `adminList`: 字符串格式的数组，里面可以写入超管的用户名，用于配置超管，如 `['minghao.yu']`。

url 中的 `{{xxx}}` 格式表示占位符，其中 `{{service}}` 会被替换成 `serviceUrl`，`{{token}}` 会被替换成获取的 token。

## 依赖安装

初始安装成功后，会提示需要进入到安装路径的 src 路径下执行 `npm install` 进行依赖安装，目前只提供淘宝源，如果有自己指定的源，可以选择 `3) other`，然后手动输入源（如输入 `http://registry.npm.xxx.org`）。

## 项目编译

安装依赖完成之后，会在 `/src` 目录下执行 `npm run build` 进行静态资源编译；如果编译出错，可以去项目路径中手动执行 `npm run build` 编译。

## 项目启动

依赖安装完成后，进入到安装路径的 src 路径下执行 `./start.sh` 进行项目启动，或者自行选择使用 pm2 等工具进行启动。

## 项目更新

在项目目录下运行

```
$ yicon update
```

运行完成之后，会先分析当前项目的变动，可以手工先保存变动的文件。之后项目会拉取最新的代码，覆盖项目内部的 `/src` 目录，并重新安装依赖和编译。

**Tips**

- 更换 Logo，请替换 `src/static/images/qiconbg.png`。


Releases.md
## v1.0.2 (2016-12-19)

### Fix bugs

- 修复大库管理无法修改名称的 bug
- 修复新建大库时，没有为管理者指定库管权限的 bug
- 修复 loading 图片大小写的 bug
- 修复下载图标时选择图标尺寸，关闭弹窗后显示尺寸与实际不符合的问题

## v1.0.1 (2016-12-16)

### Fix bugs

- 修改登录方式，改为从配置文件获取登录相关信息
- 修复替换日志记录错误 bug



es6/command/init.js
import q from 'q';
import fs from 'fs';
import path from 'path';
import chalk from 'chalk';
import inquirer from 'inquirer';
import Sequelize from 'sequelize';
import { exec } from 'child_process';

import {
  wget,
  log,
  npmPreInstall,
  buildProject
} from './utils';

const getConfig = filename => ({
  log: {
    appenders: [
      {
        category: 'normal',
        type: 'dateFile',
        filename,
        alwaysIncludePattern: true,
        pattern: '-yyyy-MM-dd.log',
      },
    ],
  },
});

let dir;

const getChoiceItem = item => item.replace(/^\d+\) /, '');

// 修复相对路径
const mixDirPath = dir => {
  const cwd = process.cwd();
  return /^\//.test(dir) ? dir : path.join(cwd, dir);
};

const $command = p => path.join(__dirname, p);
const $install = p => path.join(dir, p);

// 将 src 资源文件复制到指定目录
const copySource = async (branch) => {
  const sourceUrl = `YMFE/yicon#${branch}`;
  await q.nfcall(exec, `mkdir "${$install('/src')}"`);
  await wget(sourceUrl, $install('/src'));
  await q.nfcall(exec, `cp -r "${$install('/src')}" "${$install('.src_prev')}"`);
  await q.nfcall(exec, `cp -r "${$command('../../template/config.js')}" "${$install('/src/src')}"`);
  await q.nfcall(exec, `cp -r "${$command('../../template/start.sh')}" "${$install('/src')}"`);
  await q.nfcall(exec, `mkdir "${$install('/logs')}"`);

  log.done('项目初始化完成');
};

const getDBConfig = async (config, isDefault) => {
  if (isDefault) {
    config.model = {
      host: '127.0.0.1',
      username: 'root',
      password: '123456',
      port: '3306',
      database: 'iconfont',
      dialect: 'mysql'
    };
    return config;
  }
  log.info('输入数据库配置项，回车接受默认值');

  const questions = [
    {
      type: 'input',
      name: 'host',
      message: '请输入数据库域名或 IP 地址:',
      default: '127.0.0.1'
    },
    {
      type: 'input',
      name: 'username',
      message: '请输入数据库用户名:',
      default: 'root'
    },
    {
      type: 'input',
      name: 'password',
      message: '请输入数据库密码:',
      default: '123456'
    },
    {
      type: 'input',
      name: 'port',
      message: '请输入数据库端口号:',
      default: '3306'
    },
    {
      type: 'input',
      name: 'database',
      message: '请输入数据库名称:',
      default: 'iconfont'
    }
  ];

  config.model = await inquirer.prompt(questions);

  config.model.dialect = 'mysql';

  return config;
};

const getLoginConfig = async (config, isDefault) => {
  const ssoCasConfig = {
    authUrl: 'http://cas.example.com/cas/login?service={{service}}',
    tokenUrl: 'http://cas.example.com/serviceValidate?service={{service}}&ticket={{token}}',
    serviceUrl: 'http://app.iconfont.com',
    adminList: [],
  };
  const ldapConfig = {
    server: 'ldap://127.0.0.1:1389',
    bindDn: 'cn=root',
    baseDn: 'test',
    bindPassword: 'secret',
    searchDn: 'cn=root'
  };

  if (isDefault) {
    config.login = {};
    Object.assign(config.login, { ssoType: 'sso' }, ssoCasConfig);
    return config;
  }

  const questions = [
    {
      type: 'list',
      message: '选择登录类型',
      name: 'type',
      choices: ['1) sso', '2) cas', '3) ldap']
    }
  ];
  const { type } = await inquirer.prompt(questions);
  const ssoType = getChoiceItem(type);
  config.login = { ssoType };
  if (ssoType === 'ldap') {
    Object.assign(config.login, ldapConfig);
  } else {
    Object.assign(config.login, ssoCasConfig);
  }

  return config;
};

const getSourceConfig = async config => {
  config.source = {
    support: false,
    infoUrl: '',
    versionUrl: '',
    sourceUrl: '',
    cdn: ''
  };
  return config;
};

const writeConfigFile = async config => {
  const content = JSON.stringify(config, null, 2);
  const configFile = $install('/config.json');

  await q.nfcall(fs.writeFile, configFile, content);

  log.dim(`您的数据库配置和日志配置已写入，可以手工修改 ${chalk.yellow(configFile)} 来改变配置`);
};

const authDBConnection = async (config, isDefault) => {
  const { host, username, password, port, database } = config.model;
  const seq = new Sequelize(database, username, password, {
    port, host, logging: false, dialectOptions: { multipleStatements: true },
  });

  let initDB = false;

  try {
    await seq.authenticate();
    log.done('数据库连接成功');
    if (!isDefault) {
      const { init } = await inquirer.prompt([
        {
          type: 'confirm',
          name: 'init',
          message: '是否初始化数据库?',
          default: false
        }
      ]);
      initDB = init;
    } else {
      initDB = true;
    }

    if (initDB) {
      await importDBData(seq);
    }
  } catch (e) {
    log.error(`导入数据库失败，错误信息为：${e.message}`);
  }

  if (!initDB) {
    log.dim('如需初始化，请手工导入安装路径根目录下 sql 脚本建表');
    await q.nfcall(exec, `cp "${$command('../../template/iconfont.sql')}" "${$install('/iconfont.sql')}"`);
  }
};

const importDBData = async seq => {
  const sqlFile = $command('../../template/iconfont.sql');
  const content = await q.nfcall(fs.readFile, sqlFile);
  const sqlList = content.toString().split(';')
    .map(sql => sql.replace(/\n/g, ''))
    .filter(v => v);

  await sqlList.reduce(
    (def, sql) => def.then(() => seq.query(sql)),
    Promise.resolve()
  );
  log.done('数据导入成功');
};

const getInstallPath = async (isDefault) => {
  let _installPath = null;
  if (!isDefault) {
    const { installPath } = await inquirer.prompt({
      type: 'input',
      name: 'installPath',
      message: '请输入安装路径:',
      validate: function (value) {
        if (value) return true;
        return '安装路径不能为空';
      }
    });
    _installPath = installPath;
  } else {
    _installPath = '.';
  }
  dir = mixDirPath(_installPath.trim());

  try {
    await q.nfcall(fs.access, dir, fs.F_OK);
  } catch (err) {
    throw new Error(`${$install('/')} 路径不存在`);
  }

  const stat = await q.nfcall(fs.stat, dir);
  if (!stat.isDirectory()) {
    throw new Error(`${$install('/')} 必须是文件夹`);
  }
  const dirContent = await q.nfcall(exec, `ls ${$install('/')}`);
  if (dirContent.join('') !== '') {
    throw new Error(`请确保 ${$install('/')} 文件夹为空`)
  }

  log.dim(`正在 ${$install('/')} 路径下初始化项目...`);
};

export default async (args) => {
  try {
    const { branch = 'deploy' } = args;
    const isDefault = args && args.default;
    await getInstallPath(isDefault);
    await copySource(branch);

    const config = getConfig($install('/logs/log'));
    await getDBConfig(config, isDefault);
    await getLoginConfig(config, isDefault);
    await getSourceConfig(config);
    await writeConfigFile(config);

    await authDBConnection(config, isDefault);
    await npmPreInstall($install('/src'), $install('/install.log'), isDefault);
    await buildProject($install('/src'));

    log.done(`项目初始化成功，可以前往 ${$install('/src')} 目录下执行 ./start.sh 以 3000 端口启动服务`);
    process.exit(0);
  } catch (e) {
    log.error(e.message);
    process.exit(1);
  }
};
